### PR TITLE
feat: LID clean up

### PR DESCRIPTION
**Pull request by LexLuthr**
Fixes https://github.com/filecoin-project/boost/issues/1709

Implementation Details:
1. The process runs every 6 hours
2. It removes Completed/Slashed Boost deals, Completed Legacy Deals and TermMax reached Direct deals from LID
3. It also removed any dangling deals from LID which are not associated with any deals in SQL DB but are attributed to the miner ID

- [x] Account for FIP-45 as deal end epoch is no longer reliable

### CODE REVIEW

File: lib/pdcleaner/pdcleaner.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+		case <-ticker.C:
+			log.Infof("Starting LID clean up")
+			err := p.CleanOnce()
+			if err != nil {
+				log.Errorf("Failed to cleanup LID: %s", err)
+				continue
+			}
+			log.Info("Finished cleaning up LID")
+		case <-p.ctx.Done():
+			return
```
Comment (masih) [inline lib/pdcleaner/pdcleaner.go:99]: Log message at debug level to say that  clean up event loop has stopped?

File: lib/pdcleaner/pdcleaner.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	if err != nil {
+		return fmt.Errorf("getting complete boost deals: %w", err)
+	}
+	boostActiveDeals, err := p.dealsDB.ListActive(p.ctx)
+	if err != nil {
+		return fmt.Errorf("getting active boost deals: %w", err)
+	}
+
+	boostDeals = append(boostDeals, boostCompleteDeals...)
+	boostDeals = append(boostDeals, boostActiveDeals...)
```
Comment (masih) [inline lib/pdcleaner/pdcleaner.go:130]: Seems like `boostDeals` size is known. Instantiate it with the right size after complete and active deal slices are determined? That should result in a more efficient allocation.

File: lib/pdcleaner/pdcleaner_test.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	"github.com/ipld/go-car/v2"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPieceDirectoryCleaner(t *testing.T) {
+	req := require.New(t)
+	ctx := context.Background()
+
+	provAddr, err := address.NewIDAddress(1523)
+	require.NoError(t, err)
```
Comment (masih) [inline lib/pdcleaner/pdcleaner_test.go:37]: Use `req` consistently or remove its allocation and stick with package level exported testify library?

**Review by masih — APPROVED** [review]

File: lib/pdcleaner/pdcleaner.go [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+func NewPieceDirectoryCleaner(cfg *config.Boost) func(lc fx.Lifecycle, dealsDB *db.DealsDB, directDealsDB *db.DirectDealsDB, legacyDeals legacy.LegacyDealManager, pd *piecedirectory.PieceDirectory, full v1api.FullNode) PieceDirectoryCleanup {
+	return func(lc fx.Lifecycle, dealsDB *db.DealsDB, directDealsDB *db.DirectDealsDB, legacyDeals legacy.LegacyDealManager, pd *piecedirectory.PieceDirectory, full v1api.FullNode) PieceDirectoryCleanup {
+
+		pdc := newPDC(dealsDB, directDealsDB, legacyDeals, pd, full, cfg.LocalIndexDirectory.LidCleanupInterval)
+
+		ctx, cancel := context.WithCancel(context.Background())
+
+		lc.Append(fx.Hook{
+			OnStart: func(_ context.Context) error {
```
Comment (masih) [inline lib/pdcleaner/pdcleaner.go:51]:
```suggestion
			OnStart: func(context.Context) error {
```

File: lib/pdcleaner/pdcleaner.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+				}
+				mid, err := address.NewFromString(cfg.Wallets.Miner)
+				if err != nil {
+					return fmt.Errorf("failed to parse the miner ID %s: %w", cfg.Wallets.Miner, err)
+				}
+				pdc.miner = mid
+				go pdc.Start(ctx)
+				return nil
+			},
+			OnStop: func(_ context.Context) error {
```
Comment (masih) [inline lib/pdcleaner/pdcleaner.go:64]:
```suggestion
			OnStop: func(context.Context) error {
```

File: node/config/types.go [hunk 6]
```diff
@@ -289,6 +289,9 @@ type LocalIndexDirectoryConfig struct {
 	ServiceRPCTimeout Duration
 	// PieceDoctor runs a continuous background process to check each piece in LID for retrievability
 	EnablePieceDoctor bool
+	// Interval in hours after which LID clean up job should rerun. The cleanup entails removing indices and metadata
+	// for the expired/slashed deals. Disabled if set to 0.
+	LidCleanupInterval int
```
Comment (masih) [inline node/config/types.go:294]: I am curious why an `int` type? Would it be clearer to use `Duration` type considering it exists and that's what the ticker uses down the line?

File: lib/pdcleaner/pdcleaner.go [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+				if err != nil {
+					return err
+				}
+				if len(dd) > 0 {
+					continue
+				}
+
+				err = p.pd.RemoveDealForPiece(p.ctx, piece, deal.DealUuid)
+				if err != nil {
+					// Don't return if cleaning up a deal results in error. Try them all.
```
Comment (masih) [inline lib/pdcleaner/pdcleaner.go:255]: I am curious what happens to the dangling deals that fail to get removed?
Ditto re legacy deals above.

**Review by masih — APPROVED** [review]

File: lib/pdcleaner/pdcleaner.go (hunk 7 above)
Comment (LexLuthr) [reply lib/pdcleaner/pdcleaner.go:255]: This is to account for any DB errors. We can always retry in the later attempt. it seem inefficient to stop the cleanup for a single failure as keep that data for a little more time won't hurt the SP.

File: node/config/types.go (hunk 6 above)
Comment (LexLuthr) [reply node/config/types.go:294]: I have updated to time.Duration(). I had 2 different approaches in mind and merged the 2 which resulted in ticker with duration and var with int.

File: lib/pdcleaner/pdcleaner.go (hunk 7 above)
Comment (masih) [reply lib/pdcleaner/pdcleaner.go:290]: Proceeding seems reasonable. Though I am curious where if ever those lingering deals get cleaned up. 

Context: addressing the cleanup retry for lingering deals can certainly be done in a separate PR. Just now I need your help to understand the systems dynamics in such failure cases.

Comment (LexLuthr) [reply lib/pdcleaner/pdcleaner.go:292]: Imagine if there is issue with deleting a particular deal, it could be an ActorState or DB timeout or something else, we will not be cleaning up other deals which could have been cleaned without any issue. If such a problem comes very early in the execution then we keep holding the data we don't need. All the failures are logged and user can easily use CLI to try to remove that deal manually. This step can also help us debug the problem. 
The only exception would be we cannot CLI cannot cleanup deals with no corresponding identifier in SQL. We can handle that later if reported. 

I hope this makes sense.

File: lib/pdcleaner/pdcleaner.go [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+				if err != nil {
+					return fmt.Errorf("checking if bitfield is set: %w", err)
+				}
+				// Each claim is created with ProveCommit message. So, a sector in claim cannot be unproven.
+				// it must be either Active(Proving, Faulty, Recovering) or terminated. If bitfield is not set
+				// then sector must have been terminated. This method will also account for future change in sector numbers
+				// of a claim. Even if the sector is changed then it must be Active as this change will require a
+				// ProveCommit message
+				if !present {
+					err = p.pd.RemoveDealForPiece(p.ctx, d.PieceCID, d.ID.String())
```
Comment (rvagg) [inline lib/pdcleaner/pdcleaner.go:237]: I'm not clear on the sectors bitfield bit but the rest of this seems to me to be saying:

1. Find a list of "completed" direct deals (what's the criteria for "completed", does that mean that the deal has been handed off to the miner?)
2. Look up the allocation ID in the list of claims to see if it's been converted to a claim
3. Check that the sector has been activated, or at least is in an active state
4. if it's not active, then clean up the deal from the local piece directory

So, what are the cases this is trying to account for? There's two that I think I see:

1. "got the claim, but failed to activate the sector at the last hurdle"—which should be quite hard, but looking at the code not entirely impossible.
2. The sector was terminated but claim is dangling, as outlined in https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0045.md#sector-termination, which kind of sucks but I guess this is what makes the two-step process necessary?

Comment (LexLuthr) [reply lib/pdcleaner/pdcleaner.go:237]: Sector bitfield comes from Actor state and account for all the sectors which are active for a miner. Active sectors include the ones which are not terminated or are not yet on-chain. 
1. Complete means that the sector it AP()ed in is now in proving state.

Accounting:
1. Any deals for which we don't activate a sector. We should clean up those.
2. Any deals which were claimed but now claim has expired and sector has been terminated should be cleaned up.
3. Any deal where sector has changed should not be cleaned up. (Future reseal).
4. Any deal which creates a claim but sector goes to faulty state instead of proving should not be cleaned up. This can work as faulty sectors are part of Active sector bitfield.

2 more things to test:
1. How does Boost behaves when sector reaching faulty state instead of proving.
2. Confirm if newly proving sector (without first wdPost) are part of the Active Bitfield. If not then update code to account for those. 


To sum it up, if I am not holding a deal, I should not hold the metadata either. We can expand this logic in future if more cases or edge cases are found.

Comment (rvagg) [reply lib/pdcleaner/pdcleaner.go:237]: sounds good and reasonable

File: lib/pdcleaner/pdcleaner.go [hunk 9]
```diff
@@ -218,33 +221,89 @@ func (p *pdcleaner) CleanOnce() error {
...
 		for _, d := range completeDirectDeals {
-			// AllocationID and ClaimID should match
 			cID := verifregtypes.ClaimId(d.AllocationID)
 			c, ok := claims[cID]
+			// If claim found
 			if ok {
-				present, err := activeSectors.IsSet(uint64(c.Sector))
+				// Claim Sector number and Deal Sector number should match(regardless of how DDO works)
+				// If they don't match and older sector is removed, then we can't use the metadata
+				// This check can be removed once Curio has resealing enabled, and it can provide
```
Comment (rvagg) [inline lib/pdcleaner/pdcleaner.go:250]: 🤞 that you remember this! it'll be annoying if that lands and this is forgotten

File: lib/pdcleaner/pdcleaner.go [hunk 10]
```diff
@@ -218,33 +221,89 @@ func (p *pdcleaner) CleanOnce() error {
...
 				}
+				continue
+			}
+
+			// If no claim found
+			clientAllocs, err := verifregState.GetAllocations(d.Client)
+			if err != nil {
+				return fmt.Errorf("getting allocations for client %s: %w", d.Client, err)
+			}
+			alloc, ok := clientAllocs[d.AllocationID]
```
Comment (rvagg) [inline lib/pdcleaner/pdcleaner.go:284]: I think you can just use `verifregState.GetAllocation(d.Client, d.AllocationID)` here to make this much cheaper?

File: lib/pdcleaner/pdcleaner.go [hunk 11]
```diff
@@ -218,33 +221,89 @@ func (p *pdcleaner) CleanOnce() error {
...
+				err = p.pd.RemoveDealForPiece(p.ctx, d.PieceCID, d.ID.String())
+				if err != nil {
+					// Don't return if cleaning up a deal results in error. Try them all.
+					log.Errorf("cleaning up direct deal %s for piece %s: %s", d.ID.String(), d.PieceCID, err.Error())
+				}
+				continue
+			}
+
+			// If claim is found then we should check if it is expired. If expired, clean up the deal.
+			if alloc.Expiration < head.Height() {
```
Comment (rvagg) [inline lib/pdcleaner/pdcleaner.go:299]: these are the same logic, maybe just lump them in together, `if !ok || alloc.Expiration < head.Height() {`

File: lib/pdcleaner/pdcleaner.go (hunk 9 above)
Comment (LexLuthr) [reply lib/pdcleaner/pdcleaner.go:250]: It is not just this but the normal deal flow as well. Everything will be impacted by Curio's robustness.

File: lib/pdcleaner/pdcleaner.go (hunk 10 above)
Comment (LexLuthr) [reply lib/pdcleaner/pdcleaner.go]: Yes. Merging them.
